File: webapp/domain/nhom_sinh_vien.py
# -*- coding: utf-8 -*-
"""NHOM SINH VIEN = (CTDT, Khoa): "FTH khoa VJU2024" la MOT nhom nguoi hoc, va
mot nguoi khong the ngoi hai lop cung mot luc.

VI SAO CAN: mo hinh cu chi biet ba loai tai nguyen - GIANG VIEN (mot nguoi mot
buoi), PHONG (ltPool/labPool), va KHU VUC (khong day hai co so trong mot ngay).
Khong cho nao noi ve NGUOI HOC. Nen lich xep ra "khong trung" theo nghia cua he
thong van co the la lich ma sinh vien FTH khoa 2023 phai co mat o hai giang duong
cung tiet - do tren gio da chot trong file: 25 cap nhu vay o 12 nhom.

VI SAO LA (CTDT, Khoa) CHU KHONG PHAI RIENG KHOA: "VJU2024" khong phai mot lop
nguoi hoc - sinh vien BCSE khoa 2024 va sinh vien FTH khoa 2024 hoc hai chuong
trinh khac han, hai lop cua ho trung gio la chuyen binh thuong. Loc theo rieng
Khoa se bao hang loat vu trung khong co that; phai giao CA HAI ve moi ra dung mot
tap nguoi.

O GHEP thuoc CA HAI ve, y het domain/pham_vi.py: lop "FTH.ESAS" la lop cua nhom
FTH lan nhom ESAS (sinh vien ca hai chuong trinh cung ngoi do), "VJU2024+VJU2023"
thuoc ca hai khoa. Nen mot lop co the nam trong NHIEU nhom.

BA TRUONG HOP CUNG GIO LA DUNG Y, khong duoc bao:
  - HOC CHUNG (domain/hoc_chung.py): ca nhom la MOT buoi day vat ly, chung PHAI
    o cung o gio.
  - LOP SONG SONG cua cung mot HOC PHAN ("Xac suat - Thong ke" chia CSE3013-1 va
    CSE3013-2): sinh vien chia doi, moi nguoi hoc mot lop, trung gio la binh thuong.
  - LOP TRUC TUYEN (o "Hinh thuc" ghi truc tuyen/online): mon linh dong, o tren
    thoi khoa bieu chi de co trong danh sach dang ky chu khong phai mot buoi hoc
    co dinh. Sinh vien tu thu xep duoc nen no khong chiem cho cua mon nao.

GIOI HAN DA BIET: hai hoc phan ma sinh vien chi hoc MOT trong hai (file that:
"Tieng Nhat so cap 1" va "Tieng Nhat so cap 2" xep cung tiet cho cung mot khoa)
van bi bao la trung - he thong khong co cach nao biet dieu do tu du lieu. Bao ra
de con nguoi doc va bo qua, thay vi im lang bo sot cac vu that.

Ban SONG SINH o giao dien: frontend/src/adapters/nhomSinhVien.js - cung mot luat,
lech nhau la hop thu bao mot dang con solver ne mot dang khac.
"""

import scheduler_core as sc
from domain import bo_qua
from domain.programs import KHOA_SPLIT_RE, tach_phan


def nhom_cua_lop(data, s):
    """Cac nhom sinh vien cua mot lop -> [(ctdt, khoa)], chu thuong.

    RONG khi lop khong ghi Khoa: khong biet no cua ai thi khong ket luan gi ca -
    xem lop_chua_ghi_khoa(). Doan bua ("chac la khoa moi nhat") se sinh ra ca
    dong vu trung khong co that.

    Dung DUNG hai ham ma domain/response.py dung de sinh programParts/cohortParts
    cho giao dien, ly le nhu domain/pham_vi.py: man hinh loc ra lop nao thi solver
    phai coi lop do thuoc dung nhom ay."""
    khoa = [k.lower() for k in tach_phan(s.get("cohort"), KHOA_SPLIT_RE)]
    if not khoa:
        return []
    ctdt = [t.lower() for t in sc.section_program_names(data, s)]
    return [(c, k) for c in ctdt for k in khoa]


def nhan(nhom):
    """Nhan doc duoc cua mot nhom: ('fth', 'vju2023') -> 'FTH/VJU2023'."""
    return f"{nhom[0].upper()}/{nhom[1].upper()}"


def chiem_gio_cua_sinh_vien(data, s):
    """Lop nay co chiem thoi gian cua sinh vien khong?

    KHOA PHAI XEP -> dĩ nhiên có. Nhưng CA lop do DON VI KHAC dieu phoi ma DA CO
    GIO cung chiem: "Triet hoc Mac-Lenin" do Phong Dao tao xep Thu 2 tiet 2-5 cho
    BCSE/VJU2026 thi sinh vien khoa do dang ngoi hoc that, khoa khong duoc xep mon
    khac vao dung o ay. Do la ly do duy nhat chung phai co mat o day - chung KHONG
    vao model CP-SAT va KHONG tinh trung giang vien (xem domain/bo_qua.py).

    Lop bo qua CHUA co gio thi khong: chua ai xep thi khong co gi de ne."""
    return sc.khoa_phai_xep(s) or bo_qua.van_len_luoi(s)


def lop_chua_ghi_khoa(data):
    """Cac lop khong quy duoc ve nhom nao vi o Khoa bo trong.

    PHAI dem va noi ra tren giao dien: chung khong bi kiem, nen "0 vu trung" chi
    dung voi phan da ghi du Khoa. Im lang thi giao vu doc thanh "ca khoa sach"."""
    return [sid for sid, s in data["sections"].items()
            if chiem_gio_cua_sinh_vien(data, s) and not nhom_cua_lop(data, s)]


def cac_nhom(data, chi_xet=None):
    """{nhom: [section_id]} - dung cho ca rang buoc solver lan viec dem tren man.

    `chi_xet`: gioi han o mot tap section_id (vd rieng mot pha giai)."""
    ra = {}
    for sid, s in data["sections"].items():
        if not chiem_gio_cua_sinh_vien(data, s) or (chi_xet is not None and sid not in chi_xet):
            continue
        for g in nhom_cua_lop(data, s):
            ra.setdefault(g, []).append(sid)
    return ra


def gio_lop_khong_xep(data):
    """{section_id: slot} cua cac lop DON VI KHAC dieu phoi da co gio.

    Solver phai nhan chung duoi dang VAT CAN CO DINH (`gio_lop_pha_khac`): chung
    khong co bien trong model - khoa khong duoc dat lai gio cua chung - nhung lop
    cua khoa phai ne ra."""
    return bo_qua.gio_da_chot(data)


def hoc_truc_tuyen(s):
    """Lop nay hoc TRUC TUYEN - khong phai mot buoi hoc co dinh o mot cho, nen no
    khong giu chan sinh vien va khong chiem gio cua mon nao.

    Doc o "Hinh thuc" (cot Excel "Hinh thuc giang day").

    "LMS" KHONG duoc mien, du truoc day co. Giao vu da xac nhan: mon ghi LMS van
    la buoi hoc that, sinh vien phai co mat dung gio do. Truoc khi sua, "Tieng Anh
    B1 (LMS)" (Hinh thuc = "LMS") duoc coi la linh dong nen he thong tu do xep mon
    khac de len: tren du lieu that, "Khoa hoc toan cau va moi truong" (VJU2012-1,
    FTH/VJU2026) bi xep dung Thu 4 tiet 2-5 - trung khit lop Tieng Anh B1 cua CUNG
    nhom sinh vien do. Mien LMS go bo 175 cap khoi rang buoc.

    Con "Truc tuyen"/"Online" thi van mien: do la mon linh dong that, len thoi khoa
    bieu chi de co trong danh sach dang ky.

    "Online
Truc tiep" thi KHONG tinh: lop do CO buoi hoc that tai lop, sinh vien
    van phai co mat. Co "truc tiep" trong o la du de khong mien - huong sai an toan
    (bao thua mot vu de nguoi doc bo qua, con hon giau mat mot vu that)."""
    t = sc.bo_dau(s.get("teaching_mode"))
    if not t or "truc tiep" in t:
        return False
    return "truc tuyen" in t or "online" in t


def cung_gio_la_dung_y(data, sid_a, sid_b):
    """Hai lop nay cung gio thi KHONG phai loi - xem docstring dau file."""
    if sc.cung_nhom_hoc_chung(data, sid_a, sid_b):
        return True
    a, b = data["sections"].get(sid_a) or {}, data["sections"].get(sid_b) or {}
    # CHI MOT BEN truc tuyen la du: mon linh dong khong giu chan sinh vien o dau
    # ca, nen mon con lai muon xep vao gio do cung khong ai vuong.
    if hoc_truc_tuyen(a) or hoc_truc_tuyen(b):
        return True
    ma, mb = a.get("course_id"), b.get("course_id")
    return ma is not None and ma == mb


def cac_cap_can_ne(data, chi_xet=None):
    """Moi cap lop (a, b) ma sinh vien khong the hoc ca hai neu chung cung gio,
    kem nhan nhom de giai thich. Moi cap CHI RA MOT LAN du hai lop chung nhieu
    nhom (lop "FTH.ESAS" khoa 2023 chung ca hai nhom voi lop FTH.ESAS khac).

    Day la DAU VAO cua rang buoc mem trong scheduler_core - o do khong biet gi ve
    CTDT/Khoa, chi nhan "nhung cap nay khong duoc chong gio"."""
    ra = {}
    for g, sids in cac_nhom(data, chi_xet).items():
        ds = sorted(sids)
        for i in range(len(ds)):
            for j in range(i + 1, len(ds)):
                cap = (ds[i], ds[j])
                if cap in ra or cung_gio_la_dung_y(data, *cap):
                    continue
                ra[cap] = nhan(g)
    return ra


def cac_vu_trung(data, vi_tri, chi_xet=None):
    """Cac vu trung THAT SU tren mot bo vi tri dang co.

    `vi_tri`: {section_id: slot} - buoi nao dang nam o o nao (None/thieu = chua
    co cho, khong ket luan gi).

    Tra ve [{"nhom", "a", "b", "slot"}], sap xep on dinh de goi lai hai lan ra
    cung mot thu tu."""
    ra = []
    for (a, b), ten in sorted(cac_cap_can_ne(data, chi_xet).items()):
        sa, sb = vi_tri.get(a), vi_tri.get(b)
        if sa is None or sb is None:
            continue
        da = data["sections"][a]["duration"]
        db = data["sections"][b]["duration"]
        if not (sa + da <= sb or sb + db <= sa):
            ra.append({"nhom": ten, "a": a, "b": b, "slot": min(sa, sb)})
    return ra

File: frontend/src/adapters/nhomSinhVien.js
// NHÓM SINH VIÊN = (CTĐT, Khóa): "FTH khóa VJU2024" là MỘT nhóm người học, và
// một người không thể ngồi hai lớp cùng một lúc.
//
// Bản SONG SINH của webapp/domain/nhom_sinh_vien.py — cùng một luật gom nhóm và
// cùng một luật loại trừ. Backend dùng nó để RÀNG BUỘC solver, màn hình dùng nó
// để BÁO ra; lệch nhau là hộp thư kêu một đằng còn solver né một nẻo.
//
// VÌ SAO LÀ (CTĐT, Khóa) CHỨ KHÔNG PHẢI RIÊNG KHÓA: "VJU2024" không phải một lớp
// người học — sinh viên BCSE khóa 2024 và sinh viên FTH khóa 2024 học hai chương
// trình khác hẳn, hai lớp của họ trùng giờ là chuyện bình thường. Lọc theo riêng
// Khóa sẽ báo hàng loạt vụ trùng không có thật.
//
// Ô GHÉP thuộc CẢ HAI vế, y hệt adapters/phamVi.js: lớp "FTH.ESAS" là lớp của cả
// nhóm FTH lẫn nhóm ESAS, "VJU2024+VJU2023" thuộc cả hai khóa — nên một lớp có
// thể nằm trong nhiều nhóm.
//
// BA TRƯỜNG HỢP CÙNG GIỜ LÀ ĐÚNG Ý, không được báo:
//   - HỌC CHUNG: cả nhóm là MỘT buổi dạy vật lý, chúng PHẢI ở cùng ô giờ.
//   - LỚP SONG SONG của cùng một HỌC PHẦN (CSE3013-1 và CSE3013-2): sinh viên
//     chia đôi, mỗi người học một lớp.
//   - LỚP TRỰC TUYẾN (ô "Hình thức"): môn linh động, lên thời khóa biểu chỉ để có
//     trong danh sách đăng ký chứ không phải một buổi học cố định.
//
// GIỚI HẠN ĐÃ BIẾT: hai học phần mà sinh viên chỉ học MỘT trong hai ("Tiếng Nhật
// sơ cấp 1" và "sơ cấp 2" xếp cùng tiết cho cùng một khóa) vẫn bị báo — dữ liệu
// không nói ra điều đó. Báo để con người đọc và bỏ qua, còn hơn im lặng bỏ sót.

import { taoCungBuoi } from "./hocChung";

const thuong = (x) => String(x ?? "").toLowerCase();

// Bỏ dấu tiếng Việt để so khớp ô chữ tự do người gõ tay — bản song sinh của
// scheduler_core.bo_dau(). "đ" không tách ra khi normalize nên phải thay riêng.
const boDau = (t) =>
  String(t ?? "")
    .split(/\s+/)
    .join(" ")
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/[̀-ͯ]/g, "")
    .replace(/đ/g, "d");

/**
 * Lớp này học TRỰC TUYẾN — không phải một buổi học cố định ở một chỗ, nên nó
 * không giữ chân sinh viên và không chiếm giờ của môn nào.
 *
 * "LMS" KHÔNG được miễn, dù trước đây có: giáo vụ đã xác nhận môn ghi LMS vẫn là
 * buổi học thật, sinh viên phải có mặt đúng giờ đó. Xem bản song sinh ở
 * webapp/domain/nhom_sinh_vien.py: hoc_truc_tuyen() — hai nơi lệch nhau là hộp
 * thư báo một đằng còn solver né một đằng.
 *
 * "Online
Trực tiếp" thì KHÔNG tính — lớp đó có buổi học thật tại lớp.
 */
export function hocTrucTuyen(c) {
  const t = boDau(c?.teachingMode);
  if (!t || t.includes("truc tiep")) return false;
  return t.includes("truc tuyen") || t.includes("online");
}

// Các nhóm sinh viên của một lớp -> [["fth","vju2023"], ...].
// RỖNG khi lớp không ghi Khóa: không biết nó của ai thì không kết luận gì cả.
export function nhomCuaLop(c) {
  const khoa = (c?.cohortParts ?? []).filter(Boolean);
  if (!khoa.length) return [];
  const ctdt = (c?.programParts ?? []).filter(Boolean);
  return ctdt.flatMap((p) => khoa.map((k) => [thuong(p), thuong(k)]));
}

export function nhanNhom([ctdt, khoa]) {
  return `${String(ctdt).toUpperCase()}/${String(khoa).toUpperCase()}`;
}

/**
 * Lớp này có chiếm thời gian của sinh viên không?
 *
 * Lớp của khoa thì dĩ nhiên có. Nhưng CẢ lớp do ĐƠN VỊ KHÁC điều phối mà ĐÃ CÓ
 * GIỜ cũng chiếm: "Triết học Mác-Lênin" do Phòng Đào tạo xếp Thứ 2 tiết 2–5 cho
 * BCSE/VJU2026 thì sinh viên khóa đó đang ngồi học thật, khoa không được xếp môn
 * khác vào đúng ô ấy. Bản song sinh của webapp/domain/nhom_sinh_vien.py:
 * chiem_gio_cua_sinh_vien().
 *
 * Lớp bỏ qua CHƯA có giờ thì không: chưa ai xếp thì không có gì để né.
 */
export function chiemGioCuaSinhVien(c) {
  return !c?.boQua || (!c.timeAssumed && c.day != null);
}

// {khoá nhóm: {nhan, sectionIds}} — mọi lớp CHIẾM GIỜ của sinh viên.
export function cacNhom(data) {
  const ra = new Map();
  for (const c of data?.classes ?? []) {
    if (!chiemGioCuaSinhVien(c)) continue;
    for (const g of nhomCuaLop(c)) {
      const k = g.join("|");
      if (!ra.has(k)) ra.set(k, { nhan: nhanNhom(g), sectionIds: [] });
      ra.get(k).sectionIds.push(c.sectionId);
    }
  }
  return ra;
}

// Lớp không quy được về nhóm nào vì ô Khóa bỏ trống. PHẢI đếm và nói ra trên giao
// diện: chúng không được kiểm, nên "0 vụ trùng" chỉ đúng với phần đã ghi đủ Khóa.
export function lopChuaGhiKhoa(data) {
  return (data?.classes ?? []).filter(
    (c) => chiemGioCuaSinhVien(c) && nhomCuaLop(c).length === 0,
  );
}

function chongGio(a, da, b, db) {
  return !(a + da <= b || b + db <= a);
}

/**
 * Các vụ trùng lịch sinh viên trên một bộ buổi ĐANG NẰM THẬT trên lưới.
 *
 * @param buoi [{id, slot, duration}] — cả hai giai đoạn, đã áp bản kéo-thả chưa lưu.
 * @returns [{nhan, a, b, slot, day}] — mỗi cặp CHỈ RA MỘT LẦN dù hai lớp chung
 *   nhiều nhóm (lớp "FTH.ESAS" khóa 2023 chung cả hai nhóm với lớp FTH.ESAS khác).
 */
export function cacVuTrung(data, buoi) {
  const cungBuoi = taoCungBuoi(data);
  const monCuaLop = new Map((data?.classes ?? []).map((c) => [c.sectionId, c.courseId]));
  const trucTuyen = new Set(
    (data?.classes ?? []).filter(hocTrucTuyen).map((c) => c.sectionId),
  );
  const viTri = new Map((buoi ?? []).filter((l) => l.slot != null).map((l) => [l.id, l]));

  const daCo = new Set();
  const ra = [];
  for (const { nhan, sectionIds } of cacNhom(data).values()) {
    const ds = sectionIds.filter((sid) => viTri.has(sid)).sort((x, y) => x - y);
    for (let i = 0; i < ds.length; i++) {
      for (let j = i + 1; j < ds.length; j++) {
        const key = `${ds[i]}-${ds[j]}`;
        if (daCo.has(key)) continue;
        if (cungBuoi(ds[i], ds[j])) continue;
        // CHỈ MỘT BÊN trực tuyến là đủ — xem hocTrucTuyen().
        if (trucTuyen.has(ds[i]) || trucTuyen.has(ds[j])) continue;
        const ma = monCuaLop.get(ds[i]);
        if (ma != null && ma === monCuaLop.get(ds[j])) continue;
        const a = viTri.get(ds[i]);
        const b = viTri.get(ds[j]);
        if (!chongGio(a.slot, a.duration ?? 1, b.slot, b.duration ?? 1)) continue;
        daCo.add(key);
        ra.push({ nhan, a: ds[i], b: ds[j], slot: Math.min(a.slot, b.slot) });
      }
    }
  }
  return ra;
}
